pyne/serpent.py

```diff
@@ -370,3 +379,137 @@ def parse_det(detfile, write_py=False):
+        generated in. Then, branch names represent the keys, in the order that
+        serpent printed them. On the final dimension, the parameters returned
+        by serpent make up the remaining keys.
+
+        note: B1 is used for leakage-corrected group constants generated on
+        an infinite lattice. Don't use these values otherwise.
+
+    """
+    if isinstance(coefile, basestring):
+        f = open(coefile, 'r')
+    else:
+        f = coefile
+
+    coe = {}
+    layers = None
+
+    while True:
+
+        # this line describes branch and coe indices
+        try:
+            l = next(f).split()
+        except StopIteration:
+            break
+        _, numbranch, coeIndex, nTot, nUni = tuple([int(item) for item in l])
+
+        # this line describes what branch you're on
+        l = next(f).split()
+        # init list to hold branch names if it is still None
+        if layers is None:
+            layers = [[] for i in range(int(l[0]))]
+        for i, branch in enumerate(l[1:]):
+            # append if not seen before
+            if branch not in layers[i]:
+                layers[i].append(branch)
+        theseBranches = l[1:]
+
+        next(f)
+        l = next(f).split()
+        BUi = int(l[1])  # index of BU step
+        if BUi not in coe.keys():
+            coe[BUi] = {}
+
+        # ok, should be at a totally new set of data now,
+        # so, let's gather all params.
+        for iUni in range(nUni):
+
+            l = next(f).split()
+            uni = l[0]
+            if uni not in coe[BUi].keys():
+                coe[BUi][uni] = {}
+            nPara = int(l[1])  # num of params to follow
+
+            thisCoe = coe[BUi][uni]
+            # delve through the branches, reassigning the thisCoe reference
+            # at each successive level
+            for branch in theseBranches:
+                if branch not in thisCoe.keys():
+                    thisCoe[branch] = {}
+                thisCoe = thisCoe[branch]
+
+            names = []
+            values = []  # should all be numeric in nature
+            for i in range(nPara):
+                # build from lists to avoid frequently rehashing dicts
+                l = next(f).split()
+                name = l[0]
+                # check if B1
+                # also, yeah 'and' could be used, but bool checking
+                # is faster than string comparison, so that's why to nest
+                if not grabB1:
+                    if name[0:2] == 'B1':
+                        continue
+                value = l[2:]
+                names.append(name)
+                if len(values) == 1:
+                    values.append(float(value[0]))
+                else:
+                    values.append([float(item) for item in value])
+
+            thisCoe = coe[BUi][uni]
+            for branch in theseBranches:
+                thisCoe = thisCoe[branch]
+            thisCoe = dict(zip(names, values))
+
+    f.close()
```

Comment:
f should not be closed here if the user passed in a file handle.  It should only be closed if the user passed in a file path and we opened the handle here.